src/wagtail_personalisation/utils.py

```diff
@@ -103,17 +103,13 @@ def exclude_variants(pages):
     :return: List of pages that aren't variants
     :rtype: list
     """
-    return [
-        page for page in pages
-        if (
-            (
-                hasattr(page, 'personalisation_metadata') is False
-            ) or
-            (
-                hasattr(page, 'personalisation_metadata') and page.personalisation_metadata is None
-            ) or
-            (
-                hasattr(page, 'personalisation_metadata') and page.personalisation_metadata.is_canonical
-            )
-        )
-    ]
+
+    for page in pages:
+        if not ((hasattr(page, 'personalisation_metadata') is False
+                 ) or (hasattr(page, 'personalisation_metadata') and page.personalisation_metadata is None
+                       ) or (hasattr(page, 'personalisation_metadata') and page.personalisation_metadata.is_canonical)):
```

Comment:
This is if statement is quite difficult to read. A few suggestions I'd make:
Try to make it look a bit more like the old structure, that was well spaced and it was easy to see what related to what.
OR
Is it possible to move the closing brackets from the start of this line to the end of the previous one?
Reducing the indentation would help, does each `or` section need to be further indented like this?

Reply:
The problem with the layout of the old statement was that it caused a w504 error in flake8. The binary operator `or` cannot be at the end of start of the sentence.

Reply:
because of the braces, each `or` has to be further indented. We can just supress the error on flake8 but it seems keeping it is  something that is recommended, even though there is no clear recommendation, @KaitCrawford

Reply:
@sewagodimo can you not do the `or` lines like this then `) or (` ?
In fact, I'm reasonably sure the whole thing can be simplified to:
`if hasattr(page, 'personalisation_metadata') is not False and \`
`            page.personalisation_metadata is not None and \`
`               page.personalisation_metadata.is_canonical is not True:`
 (with some indentation that ui keeps stripping)
but whichever you prefer is fine

Reply:
thank you @KaitCrawford , your suggestion did the trick